### PR TITLE
reimplemented rotate_vectors

### PR DESCRIPTION
**Pull request by din14970**
I noticed that rotating vectors is quite slow (compared to another package that implements quaternions) due to the conversion to rotation matrices. In my opinion it kind of defeats one of the main purposes of using quaternions. I now implemented something that looks much simpler and is substantially faster (factor of 6+) using only numpy and internal functionality:

```python
v_size = 100
q_size = 1000
v = np.random.random(v_size * 3).reshape((v_size, 3))
q = quaternion.from_rotation_vector(v)

%timeit rotate_vectors(q, v)  # old
# 713 µs ± 1.95 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit rotate_vectors(q, v)  # new
# 139 µs ± 584 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

It passes all the tests but I'm not entirely sure whether the moveaxis operation will always perform as expected, would be good to doublecheck this.

### CODE REVIEW

**Comment by din14970** [comment]
@moble are you still maintaining this package and accepting PRs? I thought I would ping just in case you missed this. Sorry for bugging if you were planning on getting to it later.

**Comment by moble** [comment]
@din14970 Sorry about the slow reply.  I saw this PR and had some ideas for what to do with this, but then forgot about it.  

TL;DR: I'm running tests on [another branch](https://github.com/moble/quaternion/tree/speed_up_rotations) that makes `rotate_vectors` (A) still correct, and (B) between 2.5x to 10x faster than this PR for the cases that matter most.

---

There are a few things wrong with this PR.  First, as you suggested, it doesn't deal with multiple dimensions in the same way that the current code does — and actually errors out on lots of cases the current code handles.  But there's also a more fundamental problem with it.

> rotating vectors is quite slow ... due to the conversion to rotation matrices. In my opinion it kind of defeats one of the main purposes of using quaternions

This isn't right.  There are lots of good reasons for using quaternions, but pure speed and avoidance of rotation matrices are not among them.  In particular, while matrices are bad for composing rotations and other applications like that, they are actually very good at rotating vectors: they're numerically accurate and can be far faster than other representations, especially if BLAS gets used under the hood to do the matrix multiplications (which numpy does).

The basic problem here is that there are actually different optimal choices for rotation algorithm, depending on numerous factors.  As mentioned in the docstring, if you have just one vector to rotate with each quaternion, you're probably better off just using `R * q_v * R.conj()` — and there's no need to use a separate function in that case.  For two vectors, just do that twice.  But once you get to many more than two vectors, just in terms of number of operations, your best bet will actually be to convert to a rotation matrix.  Still use quaternions to decide what the rotations should be, but the numerical implementation of this function via matrices is faster and just as accurate.

Beyond operation counts, there are also issue like allocations, data locality, and how numpy makes more loops and copies of data than are really necessary.  Both the current code and this PR are affected by lots of that, but a statement like `quaternion.as_vector_part((np.outer(R, q_v).T*R.conj()).T).squeeze()` is probably significantly worse than the current code, because it allocates roughly the size of the output data several times over, and makes multiple loops over both the input and output data.

---

Now, having said all that, it is still true that the current code could be much faster in many cases.  It looks like I made dumb choices for how to deal with the axes, and probably should have just avoided `einsum`.  I can simplify the code to just this:
```python
    m = quaternion.as_rotation_matrix(R)
    tensordot_axis = m.ndim-2
    final_axis = tensordot_axis + (axis % v.ndim)
    return np.moveaxis(np.tensordot(m, v, axes=(-1, axis)), tensordot_axis, final_axis)
```
and get the same results, but much faster.  My measurements suggest that this code is — as expected — almost twice as slow as just using `R * q_v * R.conj()` when there's just one vector, but is 2.5x to 10x faster than this PR when there are lots of vectors to be rotated.

As mentioned above, I've put that together with some tests and added it to this package in a different branch, which I'll merge once the tests finish.

Having said all that, if the speed of rotations really impacts you that much, it might be a better idea to just implement the method suggested in the docstring, which avoids many problems associated with numpy.  For this purpose,  [`quaternionic`](https://github.com/moble/quaternionic) might be a better fit.

**Comment by din14970** [comment]
Thanks for the extensive feedback, wasn't aware of the detailed nuances. Most of our use cases concern rotating many vectors by multiple rotations, often in an "outer product" kind of way. Absolute optimal performance was not that critical, rather ease of use and properly dealing with the various dimensions. We had our own representation of orientations with quaternions, and figured it would probably be better to use this specialized package. However it was curious that rotating vectors was substantially slower than our implementation. This prompted me to investigate and make this PR.

The reason I avoided conversions to rotation matrices was because of what was written in the docs: 

> !!! note For simplicity, this function converts the input quaternion(s) to matrix form, and rotates the input vector(s) by the usual matrix multiplication.

I interpreted this to say: "implementation with rotation matrices is inferior but we do this because it is simpler". A misreading on my part then I suppose. In any case, I thought I found a rather concise way to do it only with quaternions which aimed to generalize `R * q * R.conj()`. I didn't find an example where my implementation failed so I would still be curious if you have an example? Anyway, in this context it made sense to me to make a PR as it was 6 times faster than the existing implementation.

That said, I tested your implementation and it is indeed faster as you say. On my example and hardware I'm only seeing a speed up of about 40% but I trust that it is much better than what I made. Thanks for looking into this and making the new `rotate_vectors` blazingly fast!

**Comment by moble** [comment]
> I didn't find an example where my implementation failed so I would still be curious if you have an example?

Basically, there are problems whenever the input arrays have more than one dimension, and errors whenever the input quaternion array has more than one.  So for example:

```python
R = np.normalized(quaternion.from_float_array(np.random.normal(size=(5, 7, 4,))))
v = np.random.normal(size=(11, 3,))
```

This is a 5x7 array of unit quaternions rotating 11 vectors, which fails with your code.  In general, `rotate_vectors` should accept arbitrary input array shapes, and the output should be an array of shape `R.shape + v.shape`.  This may seem like a weird requirement, but I have use cases where I frequently need `R` to have shape `(17,17,20000)`.
